src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java

```diff
@@ -176,17 +260,23 @@ private void getNextBatch() throws NamingException {
      */
     @Override
     public final boolean hasMore() throws NamingException {
+        try {
+            if (hasMoreCalled) {
+                return more;
+            }
 
-        if (hasMoreCalled) {
-            return more;
-        }
-
-        hasMoreCalled = true;
+            hasMoreCalled = true;
 
-        if (!more) {
-            return false;
-        } else {
-            return (more = hasMoreImpl());
+            if (!more) {
+                return false;
+            } else {
+                return (more = hasMoreImpl());
+            }
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
```

Comment:
The reachability fence should take 'this' as a parameter, not enumCtx. Since 'this' reachability is tracked by Cleaner and not 'enumCtx' reachability.